cpp/src/dbscan/runner.cuh

```diff
@@ -125,121 +150,201 @@ size_t run(const raft::handle_t& handle, Type_f* x, Index_ N, Index_ D,
+  for (int i = 0; i < n_batches; i++) {
+    Index_ start_vertex_id = start_row + i * batch_size;
+    Index_ n_points = min(n_owned_rows - i * batch_size, batch_size);
+    if (n_points <= 0) break;
+
+    CUML_LOG_DEBUG("- Batch %d / %ld (%ld samples)", i + 1,
+                   (unsigned long)n_batches, (unsigned long)n_points);
+
+    // i==0 -> adj and vd for batch 0 already in memory
+    if (i > 0) {
+      CUML_LOG_DEBUG("--> Computing vertex degrees");
+      ML::PUSH_RANGE("Trace::Dbscan::VertexDeg");
+      VertexDeg::run<Type_f, Index_>(handle, adj, vd, x, eps, N, D, algo_vd,
+                                     start_vertex_id, n_points, stream);
+      ML::POP_RANGE();
+    }
+    raft::update_host(&curradjlen, vd + n_points, 1, stream);
     CUDA_CHECK(cudaStreamSynchronize(stream));
-    ML::POP_RANGE();
-
-    int64_t cur_time = raft::curTimeMillis();
-    CUML_LOG_DEBUG("    |-> Took %ld ms", (cur_time - start_time));
 
     CUML_LOG_DEBUG("--> Computing adjacency graph of size %ld samples.",
                    (unsigned long)curradjlen);
-    start_time = raft::curTimeMillis();
-    // Running AdjGraph
     ML::PUSH_RANGE("Trace::Dbscan::AdjGraph");
     if (curradjlen > maxadjlen || adj_graph.data() == NULL) {
       maxadjlen = curradjlen;
       adj_graph.resize(maxadjlen, stream);
     }
-
     AdjGraph::run<Index_>(handle, adj, vd, adj_graph.data(), curradjlen,
-                          ex_scan, N, minPts, core_pts + startVertexId, algoAdj,
-                          nPoints, stream);
-
+                          ex_scan, N, algo_adj, n_points, stream);
     ML::POP_RANGE();
 
-    ML::PUSH_RANGE("Trace::Dbscan::WeakCC");
-
-    cur_time = raft::curTimeMillis();
-    CUML_LOG_DEBUG("    |-> Took %ld ms.", (cur_time - start_time));
-
     CUML_LOG_DEBUG("--> Computing connected components");
-
-    start_time = raft::curTimeMillis();
+    ML::PUSH_RANGE("Trace::Dbscan::WeakCC");
     raft::sparse::weak_cc_batched<Index_, 1024>(
-      labels, ex_scan, adj_graph.data(), curradjlen, N, startVertexId, nPoints,
-      &state, stream,
-      [core_pts, startVertexId, nPoints] __device__(Index_ global_id) {
-        return global_id < startVertexId + nPoints ? core_pts[global_id]
-                                                   : false;
+      i == 0 ? labels : labels_temp, ex_scan, adj_graph.data(), curradjlen, N,
+      start_vertex_id, n_points, &state, stream,
+      [core_pts, N] __device__(Index_ global_id) {
+        return global_id < N ? __ldg((char*)core_pts + global_id) : 0;
       });
     ML::POP_RANGE();
 
-    cur_time = raft::curTimeMillis();
-    CUML_LOG_DEBUG("    |-> Took %ld ms.", (cur_time - start_time));
+    if (i > 0) {
+      CUML_LOG_DEBUG("--> Accumulating labels");
+      ML::PUSH_RANGE("Trace::Dbscan::MergeLabels");
+      MergeLabels::run<Index_>(handle, labels, labels_temp, core_pts,
+                               work_buffer, m, N, 0, stream);
+      ML::POP_RANGE();
+    }
+  }
```

Comment:
As discussed offline, it is better to leave it as separate building blocks at the moment and document what exactly `weak_cc_batched` is doing. If we decide to integrate the merging algorithm to weakcc in the future, the following must be taken into account:

- `weak_cc_batched` would require two additional buffers for the temporary labels and the working buffer used in the merging algorithm for R
- `weak_cc_batched` would require an additional argument `bool first_batch` to know if the labels must be directly written in the accumulator array or written to a temporary buffer and merged into the accumulator
- `merge_labels` must switch to using a functor instead of a boolean mask